## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [30]:

import pandas as pd

# Load the dataset
file_path = 'client_dataset.csv'
data = pd.read_csv(file_path)

# Display the first few rows of the dataset
data.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [31]:
# View the column names in the data
column_names = data.columns
print("Column Names:", column_names)

# Use the describe function to gather basic statistics
basic_stats = data.describe()
print("Basic Statistics:\n", basic_stats)


Column Names: Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')
Basic Statistics:
           client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.2300

In [32]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [33]:
# What three item categories had the most entries?
top_categories = data['category'].value_counts().head(3)
print("Top 3 Categories with Most Entries:\n", top_categories)


Top 3 Categories with Most Entries:
 category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [34]:
# For the category with the most entries, which subcategory had the most entries?
top_category = top_categories.index[0]

# Filter data for the top category and find the top subcategory
top_subcategory = data[data['category'] == top_category]['subcategory'].value_counts().idxmax()
print(f"Subcategory with most entries in {top_category} category: {top_subcategory}")


Subcategory with most entries in consumables category: bathroom supplies


In [35]:
# Which five clients had the most entries in the data?
top_clients = data['client_id'].value_counts().head(5)
print("Top 5 Clients with Most Entries:\n", top_clients)


Top 5 Clients with Most Entries:
 client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [36]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = top_clients.index.tolist()
print("Top 5 Client IDs:", top_clients_list)


Top 5 Client IDs: [33615, 66037, 46820, 38378, 24741]


In [37]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_clients.index[0]
total_units_top_client = data[data['client_id'] == top_client_id]['qty'].sum()
print(f"Total units ordered by the top client (ID: {top_client_id}): {total_units_top_client}")


Total units ordered by the top client (ID: 33615): 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [38]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
data['subtotal'] = data['unit_price'] * data['qty']

In [39]:
data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [40]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

data['shipping_price'] =(data['unit_weight'] * data['qty']).apply(lambda x: 7 * x if x > 50 else 10 * x)
data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [41]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
data['total_price'] = (data['subtotal'] + data['shipping_price']) * 1.0925

In [42]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
data['line_cost'] = (data['unit_cost'] * data['qty']) + data['shipping_price']

data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00         5512.50  131839.076250   
1         1.49   21            0     523.95          312.90     914.258625   
2         1.68   39            6     527.28          458.64    1077.117600   
3         1.23   29            3    1056.18          356.70    1543.571400   
4        46.43   20            1    3902.00         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 22 columns]

In [43]:
# Create a column for the profit of each line using line cost and line price
data['profit'] = data['total_price'] - data['line_cost']

data.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  line_cost  \
0            1  115164.00         5512.50  131839.076250   85597.05   
1            0     523.95          312.90     914.258625     629.79   
2            6     527.28          458.64    1077.117600     765.18   
3            3    1056.18          356.70    1543.571400    1077.35   
4            1    3902.00         6500.20   11364.403500    8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [44]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    total_price = data[data['order_id'] == order_id]['total_price'].sum()
    print(f"Total price for Order ID {order_id}: {total_price}")


Total price for Order ID 2742071: 152811.897075
Total price for Order ID 2173913: 162388.7193
Total price for Order ID 6128929: 923441.24045


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [45]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients_spending = data[data['client_id'].isin(top_clients_list)].groupby('client_id')['total_price'].sum()
print("Total spending by top 5 clients:\n", top_clients_spending)



Total spending by top 5 clients:
 client_id
24741    8.226889e+07
33615    8.377309e+06
38378    1.290655e+07
46820    9.743794e+06
66037    1.025951e+07
Name: total_price, dtype: float64


In [46]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = data[data['client_id'].isin(top_clients_list)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'profit': 'sum'
}).rename(columns={
    'qty': 'Total Units Purchased',
    'shipping_price': 'Total Shipping Price',
    'total_price': 'Total Revenue',
    'profit': 'Total Profit'
}).sort_values(by='Total Profit', ascending=False)

print("Summary DataFrame:\n", summary_df)


Summary DataFrame:
            Total Units Purchased  Total Shipping Price  Total Revenue  \
client_id                                                               
24741                     239862            5126448.37   8.226889e+07   
38378                      73667            3429455.40   1.290655e+07   
66037                      43018            1395151.85   1.025951e+07   
46820                      75768            1601448.84   9.743794e+06   
33615                      64313            1828984.89   8.377309e+06   

           Total Profit  
client_id                
24741      3.657999e+07  
38378      3.271830e+06  
66037      3.255032e+06  
46820      2.736603e+06  
33615      2.201995e+06  


In [52]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 

# Define a function that converts a dollar amount to millions.
def to_millions(x):
    return x / 1_000_000

formatted_summary_df = summary_df.applymap(to_millions) 
#Apply the currency_format_millions function to only the money columns. 
formatted_summary_df.columns = [col + ' (in millions)' for col in formatted_summary_df.columns]

# Rename the columns to reflect the change in the money format. 
formatted_summary_df.rename(columns={
    'Total Units Purchased (in millions)': 'Total Units (in millions)',
    'Total Shipping Price (in millions)': 'Shipping Cost (in millions)',
    'Total Revenue (in millions)': 'Revenue (in millions)',
    'Total Profit (in millions)': 'Profit (in millions)'
}, inplace=True)

formatted_summary_df.reset_index(inplace=True)
formatted_summary_df.rename(columns={'client_id': 'Client ID'}, inplace=True)

print("Formatted Summary DataFrame:\n", formatted_summary_df)




Formatted Summary DataFrame:
    Client ID  Total Units (in millions)  Shipping Cost (in millions)  \
0      24741                   0.239862                     5.126448   
1      38378                   0.073667                     3.429455   
2      66037                   0.043018                     1.395152   
3      46820                   0.075768                     1.601449   
4      33615                   0.064313                     1.828985   

   Revenue (in millions)  Profit (in millions)  
0              82.268892             36.579992  
1              12.906551              3.271830  
2              10.259515              3.255032  
3               9.743794              2.736603  
4               8.377309              2.201995  


/var/folders/vj/hvhj49qd53n9vtf8v185v5tr0000gn/T/ipykernel_34765/96845872.py:9: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  formatted_summary_df = summary_df.applymap(to_millions)


In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = formatted_summary_df.sort_values(by='Profit (in millions)', ascending=False)
ls.display_dataframe_to_user(name="Sorted Summary", dataframe=sorted_summary_df)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995